jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java

```diff
@@ -825,314 +439,372 @@ private boolean wakeup()
+            this._interceptor = interceptor;
         }
 
-        public boolean hasContent()
+        void addContent(Content content)
         {
-            return _content.hasRemaining();
-        }
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} addContent {}", this, content);
+            if (content == null)
+                throw new AssertionError("Cannot add null content");
+            if (_consumeFailure != null)
+            {
+                content.failed(_consumeFailure);
+                return;
+            }
+            if (_rawContent != null)
+                throw new AssertionError("Cannot add new content while current one hasn't been processed");
 
-        public int remaining()
-        {
-            return _content.remaining();
+            _rawContent = content;
+            _rawContentArrived += content.remaining();
         }
 
-        public boolean isEmpty()
+        void consumeTransformedContent(Consumer<Throwable> failRawContent, Throwable failure)
         {
-            return !_content.hasRemaining();
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} consumeTransformedContent", this);
+            // start by depleting the current _transformedContent
+            if (_transformedContent != null)
+            {
+                _transformedContent.skip(_transformedContent.remaining());
+                if (_transformedContent != _rawContent)
+                    _transformedContent.succeeded();
+                _transformedContent = null;
+            }
+
+            // don't bother transforming content, directly deplete the raw one
+            if (_rawContent != null)
+            {
+                _rawContent.skip(_rawContent.remaining());
+                _rawContent.succeeded();
+                _rawContent = null;
+            }
+
+            // fail whatever other content the producer may have
+            _consumeFailure = failure;
+            failRawContent.accept(failure);
         }
 
-        @Override
-        public String toString()
+        int available(Runnable rawContentProducer)
         {
-            return String.format("Content@%x{%s}", hashCode(), BufferUtil.toDetailString(_content));
+            Content content = nextNonEmptyContent(rawContentProducer);
+            return content == null ? 0 : content.remaining();
         }
-    }
 
-    protected abstract static class State
-    {
-        public boolean blockForContent(HttpInput in) throws IOException
+        int read(Runnable rawContentProducer, byte[] b, int off, int len)
         {
-            return false;
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} read", this);
+            Content content = nextNonEmptyContent(rawContentProducer);
+            return content == null ? 0 : content.get(b, off, len);
         }
 
-        public int noContent() throws IOException
+        private Content nextNonEmptyContent(Runnable rawContentProducer)
```

Comment:
Does this still need the rawContentProducer passed in?

Reply:
Same as above.